### PR TITLE
Travis to GitHub actions and release tag

### PR DESCRIPTION
**Pull request by amartinhuertas**
Issue: https://github.com/gridap/Tutorials/issues/56

- [ ] Release tag 0.14 pending
- [ ] Did not (actually could not) check whether documentation build and deploy works with GitHub actions

Finally, I am aware that there is quite a lot of code replication in the ci.yml script. I tried to search for a solution during some minutes, without success. Perhaps there is a clever way of doing this.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+          cache-name: cache-artifacts
+        with:
+          path: ~/.julia/artifacts
+          key: ${{ runner.os }}-test-${{ env.cache-name }}-${{ hashFiles('**/Project.toml') }}
+          restore-keys: |
+            ${{ runner.os }}-test-${{ env.cache-name }}-
+            ${{ runner.os }}-test-
+            ${{ runner.os }}-
+      - uses: julia-actions/julia-buildpkg@v1
+      - run: cd test/; julia --project=.. --color=yes --check-bounds=yes runtests.jl poisson.jl validation.jl elasticity.jl
```
Comment (fverdugo) [inline .github/workflows/ci.yml:35]: `runtests.jl` runs all tests. Not needed here, right?

idem in the other jobs.

**Review by fverdugo — CHANGES_REQUESTED** [review]
Thanks @amartinhuertas !

See my comment below.

By the way, some checks have been cancelled. Do you know, why?

**Comment by amartinhuertas** [comment]
> Thanks @amartinhuertas !
> 
> See my comment below.
> 
> By the way, some checks have been cancelled. Do you know why?

I cancelled them manually.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (amartinhuertas) [reply .github/workflows/ci.yml:35]: > `runtests.jl` runs all tests. Not needed here, right?
> 
> idem in the other jobs.

Yes, you are right. I thought it was working when I went to bed yesterday, but it wasn't. 

Solved in ca0ecccdd02e67c30cb5bc63f975b2cf2b75735b. Note that the idea now is that `runtests.jl` (optionally) gets command-line arguments with the particular set of tutorial filenames to be executed.  

With parallel jobs, although there is a little bit of job unbalance (I tried to balance the load manually by looking at the computational times), now it takes around 7 minutes to execute all tutorials. Not that bad ...

**Review by fverdugo — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (fverdugo) [reply .github/workflows/ci.yml:35]: > 7 minutes to execute all tutorials.

Great job!

**Comment by fverdugo** [comment]
@amartinhuertas I was going to merge, but just realized that we don't have set up the DOCUMENTER_KEY. So documentation will not work.

The idea is quite simple. Just in case this is the first time you do this:

- You have to generate locally a special public/private key pair with documenter tools: https://juliadocs.github.io/Documenter.jl/stable/man/hosting/#travis-ssh
- The public key goes in the deploy keys section of the repo (do not forget to give it write acces): https://github.com/gridap/Tutorials/settings/keys
- The private key goes to the secrets section of the repo (never `echo $DOCUMENTER_KEY` in a workflow): https://github.com/gridap/Tutorials/settings/secrets/actions

**Review by fverdugo — CHANGES_REQUESTED** [review]
See my comment on DOCUMENTER_KEY above. Thanks!

**Comment by amartinhuertas** [comment]
> See my comment on DOCUMENTER_KEY above. Thanks!

Thanks for the instructiions ... Done! My concern above in the form of a pending task was more than justified! There is always a little nudge of extra work to be performed !  Recall that we still have to add the pending release tag. To which commit do we add it? Better once this branch is merged to master, right?

**Comment by fverdugo** [comment]
>To which commit do we add it? Better once this branch is merged to master, right?

Yes! Once in master, in a commit for which all tests passed and the documentation was successfully generated.

**Review by fverdugo — APPROVED** [review]

**Comment by fverdugo** [comment]
Here we go!

**Comment by fverdugo** [comment]
> There is always a little nudge of extra work to be performed ! 

Yes! E.g., the badge in the README.md file.

**Comment by amartinhuertas** [comment]
> > There is always a little nudge of extra work to be performed !
> 
> Yes! E.g., the badge in the README.md file.

Ok ... I will do it. Anything else we might be missing?

**Comment by fverdugo** [comment]
> Anything else we might be missing?

Yes. I am not sure if the CI will be triggered when pushing a new tag.

**Comment by amartinhuertas** [comment]
> > Anything else we might be missing?
> 
> Yes. I am not sure if the CI will be triggered when pushing a new tag.

I would say that it won't (if and only if events are non-overlapping): there is an special event for such particular purpose: https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#create

**Comment by amartinhuertas** [comment]
why do want to know? What is your concern?

**Comment by fverdugo** [comment]
> why do want to know? What is your concern?
We want that documenter generates a new version of the documentation when pushing a new tag. For registered packages, I assume that TagBot does the job.  But Tutorials is not a registered package...

**Comment by amartinhuertas** [comment]
>We want that documenter generates a new version of the documentation when pushing a new tag. 

For my understanding ... why do we need that ? 

Do u want the tests to be also be re-rerun when pushing a new tag? I think that's not necessary, right? Thus we can either add an if statement to the tests to skipe them if event == create, or to move the documentation generation to a separate workflow. Which do u think it is better?

**Comment by fverdugo** [comment]
Run test also please!

Just in case somebody pushes a tag "a la brava"  😆

**Pull request merged**